src/monad/monad.ml

```diff
@@ -0,0 +1,51 @@
+module State = struct
+  type ('s, 'a) t = State of ('s -> 's * 'a)
+
+  let return x = State (fun s ->
+      (s, x)
+    )
+
+  let fmap f (State g) = State (fun s ->
+      let (s', x) = g s in
+      (s', f x)
+    )
+
+  let bind (State g) f = State (fun s ->
+      let (s', x) = g s in
+      let State g' = f x in
+      g' s'
+    )
+
+  let seq x y = bind x (fun _ -> y)
+
+  let get () = State (fun s ->
+      (s, s)
+    )
+
+  let set s = State (fun _ ->
+      (s, ())
+    )
+
+  let modify f =
+    bind (fmap f @@ get ()) set
+
+  let run (State g) s =
+    g s
+
+  let exec m s = snd (run m s)
+
+  let rec fold_leftM f acc l =
+    match l with
+    | [] -> return acc
+    | x::xs ->
+      bind (f acc x) (fun w -> fold_leftM f w xs)
```

Comment:
This function is not tailrec while potently used on large lists.

Reply:
Ocamlc manages to optimize the recursive call because it is enclosed in the continuation (confirmed experimentally, and by looking at the output of the intermediate representation of ocamlopt).

Instead of rewriting the function using the CPS form for instance which would make the code more cumbersome, wouldn't it be simpler just to put an annotation to guarantee the optimization and change that if needed ?

```ocaml
  let rec fold_leftM f acc l =
    match l with
    | [] -> return acc
    | x::xs ->
      bind (f acc x) (fun w -> (fold_leftM [@tailcall]) f w xs) 
```